File: src/com/vm/ClassFile.java
package com.vm;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.*;

public class ClassFile {
    private static final List<Byte> res = new ArrayList<>();

    public static void main(String[] args) {
        appendU4(0XCAFE_BABEL);
        appendU4(52);

        //共计 26 个常量
        appendU2(26);
        //#1
        appendUtf8("com/TestCode");
        //#2
        appendUtf8("java/lang/Object");
        //#3
        appendUtf8("java/lang/System");
        //#4
        appendUtf8("java/io/PrintStream");
        //#5
        appendUtf8("Ljava/io/PrintStream;");
        //#6
        appendUtf8("out");
        //#7
        appendUtf8("<init>");
        //#8
        appendUtf8("()V");
        //#9
        appendUtf8("print");
        //#10
        appendUtf8("(Ljava/lang/String;)V");
        //#11
        appendUtf8("main");
        //#12
        appendUtf8("([Ljava/lang/String;)V");
        //#13
        appendUtf8("Hello Class File");
        //#14
        appendUtf8("Code");

        //#15
        appendClassInfo(1);
        //#16
        appendClassInfo(2);
        //#17
        appendClassInfo(3);
        //#18
        appendClassInfo(4);

        //#19: init ()void
        appendNameAndTypeInfo(7, 8);
        //#20: out Ljava/io/PrintStream;
        appendNameAndTypeInfo(6, 5);
        //#21: print (String)void
        appendNameAndTypeInfo(9, 10);

        //#22: Object.init()v
        appendMethodRefInfo(16, 19);
        //#23: PrintStream.print(String)v
        appendMethodRefInfo(18, 21);
        //#24: System.out:PrintStream;
        appendFieldRefInfo(17, 20);

        //#25: "Hello Class File"
        appendStringInfo(13);

        appendU2(0x21);
        //类名TestCode
        appendU2(15);
        //父类Object
        appendU2(16);
        //接口数量:0
        appendU2(0);

        //跳过字段信息
        appendU2(0);

        //方法个数:2
        appendU2(2);
        // public void init ()
        appendMethod(0x0001, 7, 8);
        appendU2(0);

        //public static void main (String)
        appendMethod(0x0009, 11, 12);
        appendU2(0);

        //属性信息
        appendU2(0);

        print();
        writeClassFile();
    }


    private static void appendU1(int u1) {
        appendU1(Short.parseShort(String.valueOf(u1)));
    }

    private static void appendU1(short u1) {
        if ((u1 & 0XFF00) != 0) {
            throw new RuntimeException("appendU1 error,value:" + u1);
        }
        res.add((byte) (u1 & 0xFF));
    }

    private static void appendU2(int u2) {
        if ((u2 >> 16) != 0) {
            throw new RuntimeException("appendU2 error,value:" + u2);
        }
        res.add((byte) (u2 >> 8 & 0xFF));
        res.add((byte) (u2 & 0x00FF));
    }

    private static void appendU4(long u4) {
        if ((u4 >> 32) != 0) {
            throw new RuntimeException("appendU4 error,value:" + u4);
        }
        res.add((byte) (u4 >> 24 & 0xFF));
        res.add((byte) (u4 >> 16 & 0xFF));
        res.add((byte) (u4 >> 8 & 0xFF));
        res.add((byte) (u4 >> 0 & 0xFF));
    }


    //添加一个字符串常量
    private static void appendUtf8(String string) {
        appendU1(1);
        appendU2(string.length());
        byte[] bytes = string.getBytes();
        for (byte b : bytes) {
            res.add(b);
        }
    }

    private static void appendClassInfo(int index) {
        appendU1(7);
        appendU2(index);
    }

    private static void appendNameAndTypeInfo(int nameIndex, int descriptorIndex) {
        appendU1(12);
        appendU2(nameIndex);
        appendU2(descriptorIndex);
    }

    private static void appendMethodRefInfo(int classIndex, int nameTypeIndex) {
        appendU1(10);
        appendU2(classIndex);
        appendU2(nameTypeIndex);
    }

    private static void appendFieldRefInfo(int classIndex, int nameTypeIndex) {
        appendU1(9);
        appendU2(classIndex);
        appendU2(nameTypeIndex);
    }

    private static void appendStringInfo(int stringIndex) {
        appendU1(8);
        appendU2(stringIndex);
    }

    private static void appendMethod(int accessFlags, int methodName, int descriptor) {
        appendU2(accessFlags);
        appendU2(methodName);
        appendU2(descriptor);
    }


    private static void print() {
        for (int i = 0; i < res.size(); i++) {
            Byte b = res.get(i);
            System.out.printf("%02X", b);
            if (i % 4 == 3) {
                System.out.print(" ");
            }
        }
    }

    private static void writeClassFile() {
        try {
            File file = new File("./TestCode.class");
            file.delete();
            FileOutputStream fos = new FileOutputStream(file);
            byte[] bytes = new byte[res.size()];
            for (int i = 0; i < res.size(); i++) {
                bytes[i] = res.get(i);
            }
            fos.write(bytes);
            System.out.print("\n数据写入完毕");
            fos.close();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/com/vm/ClassFile.kt
package com.vm


import com.链表.Code
import java.math.BigInteger

val CONSTANT_Utf8_info = 1

//类或接口的全限定名
val CONSTANT_Class_info = 7

val CONSTNAT_Methodref_info = 10

//方法签名: 方法名+参数
val CONSTANT_NameAndType_info = 12

val sb = StringBuilder()
fun main() {
    //魔数
    appendU4("CAFEBABE")

    //副版本号
    appendU2(0)
    //主版本号
    appendU2(52)


    appendU2(16)

    appendReferenceConstant(CONSTNAT_Methodref_info, 3, 13)


    appendReferenceConstant(CONSTANT_Class_info, 14)
    appendReferenceConstant(CONSTANT_Class_info, 15)



    appendUtf8("<init>")
    appendUtf8("()V")
    appendUtf8("Code")
    appendUtf8("LineNumberTable")
    appendUtf8("LocalVariableTable")
    appendUtf8("this")
    appendUtf8("Lcom/TT;")
    appendUtf8("SourceFile")
    appendUtf8("TT.java")

    //init 方法 void init()
    appendReferenceConstant(CONSTANT_NameAndType_info, 4, 5)
    appendUtf8("com/TT")
    appendUtf8("java/lang/Object")

    //访问标志位
    appendU4(0x0021)
    //类名
    appendU4(0x0002)
    //父类
    appendU4(0x0003)

    // 接口数量
    appendU4(0x0000)

    // filed 数量
    appendU4(0x0000)

    // 方法数量, 一个 init 方法
    appendU4(0x0001)
    // public void init()
    appendMethod(0x0001, 4, 5)
    appendMethodCode(6, 47)
    appendMethodCodeTable(1, 1, 5)
    appendInstruction(Code.aload_0)
    appendInstruction(Code.invokespecial)
    appendU2(1)
    //exception_table_length
    appendU2(0)
    //属性表个数
    appendU2(2)
    appendLineNumberTable(7, 6, 1)
    appendLineNumber(0, 3)

    appendInstruction(Code.return_)

    print()

}


fun appendU1(value: Int) {
    if (value !in 0..0XFF) {
        throw Exception("appendU1 error:${value}")
    }
    sb.append(String.format("%02x", value))
}


//添加一个字符串常量
fun appendUtf8(value: String) {
    appendU1(CONSTANT_Utf8_info)
    appendU2(value.length)
    val bytes: ByteArray = value.toByteArray()
    val hexString = BigInteger(1, bytes).toString(16)
    sb.append(hexString)
}

//添加引用类型的常量
fun appendReferenceConstant(type: Int, index: Int) {
    appendU1(type)
    appendU2(index)
}

fun appendReferenceConstant(type: Int, index: Int, index2: Int) {
    appendU1(type)
    appendU2(index)
    appendU2(index2)
}


//访问标志 方法名 方法出参入参
fun appendMethod(accessFlags: Int, methodName: Int, descriptor: Int) {
    appendU2(accessFlags)
    appendU2(methodName)
    appendU2(descriptor)
}

//属性名称
fun appendMethodCode(attributeName: Int, length: Long) {
    appendU2(attributeName)
    appendU4(length)
}

//属性名称
fun appendMethodCodeTable(maxStack: Int, maxLocals: Int, codeLength: Long) {
    appendU2(maxStack)
    appendU2(maxLocals)
    appendU4(codeLength)
}

fun appendInstruction(code: Int) {
    appendU1(code)
}

fun appendLineNumberTable(attributeName: Int, attributeLength: Long, tableLength: Int) {
    appendU2(attributeName)
    appendU4(attributeLength)
    appendU2(tableLength)
}

fun appendLineNumber(startPc: Int, lineNum: Int) {
    appendU2(startPc)
    appendU2(lineNum)
}


fun appendU2(value: Int) {
    if (value !in 0..0XFFFF) {
        throw Exception("appendU2 error:${value}")
    }
    sb.append(String.format("%04x", value))
}


fun appendU2(value: String) {
    if (value.length != 4) {
        throw Exception("appendU2 error:${value}")
    }
    sb.append(value)
}


fun appendU4(num: Long) {
    sb.append(String.format("%08x", num))
}

fun appendU4(value: String) {
    if (value.length != 8) {
        throw Exception("appendU4 error:${value}")
    }
    sb.append(value)
}


fun print() {
    val sb = sb.toString().toCharArray()
    var count = 0
    for (v in sb) {
        print(v)
        count++
        if (count % 8 == 0) {
            print(" ")
        }
    }
}
